File: 2025/day1.py
def read():    
    return map(lambda x: int(x[1: len(x)+1]) if x[0] == "R" else -int(x[1: len(x)+1]) ,open('./inputs/day1.txt', 'r').read().split('\n'))

def part1():
    count = 0
    rotation = 50
    for i in read():
        rotation =  rotation + i % 100
        count += ((rotation // 100) * 100 == rotation)


    print(count)

def part2():
    count = 0
    rotation = 50
    for i in read():
        dir = 1 if i > 0 else -1
        for _ in range(abs(i)):
            rotation += dir
            if rotation == 100:
                rotation = 0
            if rotation == -1:
                rotation =99
            if rotation == 0:
                count +=1

    print(count)

File: 2024/rust/day2/day1/day_1.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};

fn part1() -> io::Result<()> {
    let mut list1: Vec<i32> = Vec::new();
    let mut list2: Vec<i32> = Vec::new();

    let file = File::open("day1/input.txt")?;
    let reader = io::BufReader::new(file);
    for line in reader.lines() {
        let line = line?;
        let mut iter = line.split_whitespace();
        list1.push(iter.next().unwrap().parse::<i32>().unwrap());
        list2.push(iter.next().unwrap().parse::<i32>().unwrap());
    }

    list1.sort();
    list2.sort();

    println!(
        "Result part 1: {}",
        list1
            .iter()
            .zip(list2.iter())
            .map(|(x, y)| (x - y).abs())
            .sum::<i32>()
    );

    Ok(())
}

fn part2() -> io::Result<()> {
    let mut list1: Vec<i32> = Vec::new();
    let mut count: HashMap<i32, i32> = HashMap::new();

    let file = File::open("day1/input.txt")?;
    let reader = io::BufReader::new(file);
    for line in reader.lines() {
        let line = line?;
        let mut iter = line.split_whitespace();
        list1.push(iter.next().unwrap().parse::<i32>().unwrap());

        count
            .entry(iter.next().unwrap().parse::<i32>().unwrap())
            .and_modify(|e| *e += 1)
            .or_insert(1);
    }

    println!(
        "Result part 2: {}",
        list1
            .iter()
            .map(|x| x * count.get(x).unwrap_or(&0))
            .sum::<i32>()
    );

    Ok(())
}

fn main() -> io::Result<()> {
    part1();
    part2();
    Ok(())
}
